### DISCUSSION TITLE
_next/static/1589126454533/pages/limited.js net::ERR_ABORTED 404 (Not Found)

### DISCUSSION THREAD

**Discussion by harpreetkhalsagtbit** [vercel/next.js#12684] [Help]
Code is build with `npm run build`

my folder structure:
`/pages/[url].js`

I am using link as

```js

<Link href="/[url]" as={`/${service.url}`}><a>{service.name}</a></Link>

```

**Comment by jamesmosier** [comment]
I created a minimal project with this, ran `yarn build && yarn start` but did not see that issue.

In the past though I've come across that when there are other `<Link>` component on the page which contain invalid values (for example `<Link href="#">` i've seen cause issues with other links on a page). 

Are there any other `<Link>`s? In the nav? footer? elsewhere?

**Comment by harpreetkhalsagtbit** [comment]
@jamesmosier 
Thanks for putting in the effort. Indeed, yo put me in the right direction. Turns out that, I was using `<Link>` for almost every `anchor tag`. Now I got to know that only dynamic route should be created with `<Link>`. Correct me If I am wrong.

That problem is fixed
What I do is:
1. Instead of using `<Link><a></a></Link>` for external urls I used `<a></a>`
2. Converted `<Link href="#"</Link>` to  `<a href="#"</a>`
3. Converted `<Link href=""</Link>` to  `<a href=""</a>`

Now, I am in another issue
`<Link href="/[url]" as={`/${service.url}`}><a>{service.name}</a></Link>` is not working. The URL gets updated but it does not change the page.

**Edit:**
OK, I am using SSR - **getStaticPaths** and **getStaticProps** and updated to `9.4`

`<Link href="/[url]" as={`/${service.url}`}>` was not working,  I changed it to `<a href={`/${service.url}`}{service.name}</a>` and now urls are working fine.

I guess, Is it because after build they are not dynamic anymore they are already created at the server.

**Reply by jamesmosier** [reply]
I would still expect your `<Link>` component to work. See the [with-static-export](https://github.com/zeit/next.js/blob/canary/examples/with-static-export/components/post.js#L7) example in which it links to a page in `pages/post/[id].js` which uses both `getStaticPaths` and `getStaticProps`

**Comment by aaron-fuleki** [comment]
I think I'm seeing something similar after upgrading a project to 9.4, but with my `api` endpoints. Links work fine in local dev, but once deployed, they 404. 

I see 404's for things like: `https://my-app.example.com/_next/static/<BUILD_HASH>/pages/api/login.js`

When I ssh into the running instance, the file's clearly not there. Upon poking around, I see that it _is_ in `server/static/<BUILD_HASH>/pages/api/login.js`. So it's there, just not where the server process is serving it from, based on however the path was written.

This worked fine before upgrading. For now I'm just ripping out `next/link`, since I didn't really use it much anyways, and using plain-old `a` tags. However, I would like to understand better how it decides what compiled output goes/maps to where.

Did I miss a doc update? Y'all move so fast, it's hard to keep up! 🏃

**Reply by jamesmosier** [reply]
Hi @aaron-fuleki. It is hard to say without seeing your code, but are you linking to an API endpoint explicitly?

**Reply by aaron-fuleki** [reply]
I was wrapping Material-UI `<Button>` components with a Next `<Link>`, like so:

```
<Link href="/api/logout"> 
  <Button color="inherit" className={classes.logoutButton}>
    Logout
  </Button>
</Link>
```
That should refer to `pages/api/logout.ts`, which is pretty much taken verbatim from the official Auth0 example, but converted to Typescript. It worked as expected in < 9.4, but the build process clearly maps to the wrong output artifact in 9.4+.

Converting everything to `<a>` tags like @harpreetkhalsagtbit suggested works fine. 

What was super weird was that you'd see 404's in the network panel even before clicking the button, like the build process thought the api code was a client bundle fragment that it should lazy load. That seems very wrong; I _never_ want server-side API code leaking to the client.

I'm on a deadline now, but will try to setup a clean repro when I come up for air.

EDIT: should I have sandwiched an `<a>` between those two? I may not be up on the very latest best practice, there.

**Reply by jamesmosier** [reply]
`next/link` must have an `<a>` as a child, not a `button` (assuming `<Button>` is a button element) since a `button` cannot have an `href` attribute.

If the child of `<Link>` is a custom component [you can use `passHref`](https://nextjs.org/docs/api-reference/next/link#if-the-child-is-a-custom-component-that-wraps-an-a-tag) (for example, a styled components `a`).
